⚙️ 스타일 설정

Legitgoons opened the pull request:
## 작업 이유

Webview의 font, color등 style 설정
Layout을 위한 FNB 구현
<br/>

## 작업 사항

### 1️⃣ font설정
- 디자인팀에서 정의해준 font에 알맞게 font를 설정하였습니다.
![image](https://github.com/CollaBu/pennyway-client-webview/assets/101088491/7692b421-6739-43ea-8b63-e91be26757e6)
- mixin을 사용하여 통일된 형식의 class로 사용할 수 있도록 했습니다.
```scss
@mixin font-style($weight, $size, $line-height){
/...
}
.h1bold { @include font-style(bold, 1.5, 130%); }
```
- font 사이즈 이름 + weight의 첫 글자를 따서 className으로 설정했습니다.
```ts
<h1 className='h1bold '>이렇게 하면 h1bold 설정이 적용됩니다.</h1>
```
- 용량을 줄이고 빠르게 폰트를 가져올 수 있도록 subset font를 사용했습니다.
- 용량이 50%가량 더 적은 woff2를 기본 폰트로 하고, woff를 fallback font로 지정했습니다.

### 2️⃣ color 설정
- 디자인팀에서 정의해준 color에 알맞게 color를 설정하였습니다.
![image](https://github.com/CollaBu/pennyway-client-webview/assets/101088491/91624359-063d-4230-8b5d-951719eea93f)
- 위의 색상 중, feed에서 사용되는 Gray1(가장 밝은 색)~7(가장 어두운 색)을 g1부터 g7로 설정했습니다.
- Mint1(가장 밝은 색)~3(가장 어두운 색)은 m1부터 m3로 설정했습니다.
- 이 외의 색은 feed에서 사용되지 않기에, 설정하지 않았습니다.

### 3️⃣ Layout 설정
![image](https://github.com/CollaBu/pennyway-client-webview/assets/101088491/77e2c76f-8bac-44e2-be50-b650fa41eaa6)
- iOS에서 Navbar를 만들 것을 고려해 임시로 Navbar를 만들어 하단 공간을 점유하도록 했습니다.
- user-scalable=no를 viewport meta태그에 추가해, 의도치 않은 확대가 일어나지 않게했습니다.
```html
    <meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=no" />
```

### 기타
- 전처리기인 sass의 특성을 고려하여 devDependencies로 재설치하였습니다.

<br/>

## 리뷰어가 중점적으로 확인해야 하는 부분

- [x] fonts의 class들이 제대로 구현되어 있나요?
- [x] colors 변수들이 제대로 설정되어 있나요?
- [x] Layout이 제대로 설정되어 있나요?

<br/>

## 발견한 이슈
iOS의 경우 폰트 사이즈가 16px 이하인 상태에서 input태그를 누르면 자동으로 화면이 확대되게 됩니다. 이를 방지하고자 'user-scalable=no'를 설정했습니다. 좀 더 자세한 사항은 https://im-designloper.tistory.com/49 이 링크를 참조해주세요.

Legitgoons committed:
feat: Pretendard font 설치

용량을 줄이고 빠르게 폰트를 가져올 수 있도록 subset font를 사용했습니다.

Legitgoons committed:
feat: mixin을 사용해 font class 설정

용량이 50%가량 더 적은 woff2를 기본 폰트로 하고, woff를 fallback font로 지정했습니다

Legitgoons committed:
feat: color 변수 설정

Legitgoons committed:
feat: 임시 Navbar 컴포넌트 구현

Legitgoons committed:
feat: iOS에서 Navbar를 구현할 공간을 고려하여 RootLayout에 Navbar 추가

Legitgoons committed:
feat: 의도하지 않은 확대를 막기위해 user-scalable=no 설정

Legitgoons committed:
feat: 전처리기인 sass의 특성을 고려해 devDependencies로 재설치

Legitgoons committed:
fix: naming rule에 알맞게 font파일들 이름 변경

Legitgoons committed:
fix: 호출 경로 수정 및 mixin 매개변수에 name 제거

BangDori commented on index.html:
```diff
@@ -3,7 +3,7 @@
     <meta charset="UTF-8" />
     <link rel="icon" type="image/svg+xml" href="/icon.svg" />
-    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
+    <meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=no" />
```
P3: 의도치 않은 확대가 일어나도 큰 문제가 없다고 생각하는데, 혹시 설정하신 이유가 있으신가요?

Legitgoons replied:
iOS의 경우 폰트 사이즈가 16px 이하인 상태에서 input태그를 누르면 자동으로 화면이 확대되게 됩니다. 이를 방지하고자 설정했습니다. 좀 더 자세한 사항은 https://im-designloper.tistory.com/49 이 링크를 참조하시면 좋을 것 같습니다.

하지만 이 부분은 기획을 어떻게 할 지 여부에 따라 변동 될 수 있는 사항이니 PM님에게 문의드려보겠습니다.

BangDori replied:
확실히 이 부분에 대해서는 팀원들과 함께 이야기를 나누어보면 좋을 것 같네요!

새로운 걸 알게 되었네요. 감사합니다! ㅎㅎ

BangDori commented on package.json:
P5: 이 부분은 실제로 npm에 등록된 [sass](https://www.npmjs.com/package/sass) Usage를 확인해보니, devDependencies에 등록하는 걸 권장하고 있더라구요!! 변경해주셔서 감사합니다!

BangDori commented on src/shared/styles/navber.scss:
P4: `Navbar.tsx` 컴포넌트에서 사용되는 스타일이라면 컴포넌트와의 통일성을 위해 Navbar.scss가 더 괜찮지 않을까요?

Legitgoons replied:
일단 오타가 있었네요 ㅎㅎ;; 
- 다른 스타일 시트 파일들도 카멜 케이스로 작성했기에 카멜 케이스로 작성하는 것이 일관성면에서 낫다고 생각합니다.
- 다만 아직 스타일 시트의 경우 아직 명확한 네이밍 룰을 정하지 않았네요. 이 부분부터 결정해야할듯합니다.

BangDori replied:
스타일 시트의 사용 용도를 생각해보면, 제가 생각하기에는 다음과 같이 분리될 수 있다고 생각합니다.

- 컴포넌트에 대한 스타일 시트 ex) `Component.scss` - 파스칼 케이스
- 공용 스타일 시트 ex) `colors.scss` - 카멜 케이스

그러므로 스타일 시트의 용도를 컨벤션을 통해서 한층 더 빠르게 이해하고자, 컴포넌트에 대한 스타일 시트는 컴포넌트의 컨벤션을 따르고, 공용 스타일 시트는 카멜케이스를 따르는 방법을 제시해봅니다!

Legitgoons replied:
컴포넌트에 종속되는 스타일 시트라면 이렇게 하는 게 더 적절해 보입니다. 수정하고, Naming rule에 작성하도록 하겠습니다!

BangDori commented on src/shared/ui/Navbar.tsx:
P2: 공용 컴포넌트의 경우에는 Navbar 외에도, 많은 공용 컴포넌트가 `./shared/ui/`에 들어올 수 있다고 생각합니다! 그렇기 때문에, `shared/ui/{공용 컴포넌트}` 으로 설정해두고 내부에 스타일 시트와 컴포넌트를 함께 위치하는 것이 더 가독성이 좋을 것 같습니다!

```md
📦src
 ┣ 📂shared 공유 폴더
 ┃ ┣ 📂styles 전역 스타일 설정
 ┃ ┃ ┣ 📜colors.scss
 ┃ ┃ ┣ 📜fonts.scss
 ┃ ┗ 📂ui 재사용 가능한 컴포넌트 / 공용 컴포넌트
 ┃ ┃ ┣ 📂navigation 네비게이션 UI
 ┃ ┃ ┃ ┣ 📜GNB.tsx
 ┃ ┃ ┃ ┣ 📜GNB.scss
```

실제로 FSD Architecture를 기반으로 구현한 sudoku-effector 프로젝트의 [shared/ui](https://github.com/Shiyan7/sudoku-effector/tree/master/src/shared/ui) 폴더를 확인해본 결과, 아래와 같이 구성되어 있습니다.

<img src="https://github.com/CollaBu/pennyway-client-webview/assets/44726494/0c21810f-5f72-4091-991d-be76d88d5015" width="360" />

Legitgoons replied:
해당 방향성이 맞는 것 같습니다. 이렇게 수정하도록 하겠습니다!

BangDori commented on src/shared/ui/Navbar.tsx:
```diff
@@ -0,0 +1,5 @@
+import '../styles/navber.scss'
+
+export default function Navbar() {
```
P4: 현재 Navbar 컴포넌트의 경우, 컴포넌트의 이름만으로는 어떤 역할을 하는? 어떤 위치에 존재하는 네비게이션 바인지 유추하기 힘들다고 생각합니다! 그러므로 GNB, LNB, SNB와 같은 네이밍을 사용해보는 것도 좋을 것 같습니다!

Legitgoons replied:
상단 Navbar도 있어 오해의 소지가 있어보입니다. FNB로 컴포넌트 명을 수정하겠습니다.

BangDori commented on src/shared/styles/colors.scss:
P4: colors에 설정해주신 스타일 변수(`g1, g2, g3, ...`)의 네이밍을 보고선 어떤 변수에 대한 색상인지가 크게 와닿지 않는 것 같습니다..! 이렇게 명명하신 이유가 궁금합니다!

Legitgoons replied:
스타일 변수는 알파벳 + 넘버링의 구조로 만들었습니다.
- 알파벳: feed에서는 gray, mint 두 가지만 사용하기에 앞 글자만 표기해도 헷갈리지 않을 것이라 생각했습니다.
- 넘버링: 디자인 팀이 figma에서 색상을 넘버링으로 처리했기에 일관적으로 넘버링을 달아주는 것이 작업하는데 효율이 날 것이라고 생각했습니다.

BangDori replied:
프로젝트를 설계하는 현 시점에서는 이를 무리하는데 큰 지장이 없지만, 추후에 프로젝트를 다시 살펴보는 시점에서 혼란스러울 수도 있다고 생각합니다!

- g1: gray01, green01, ...

물론 피그마를 살펴보면 해당 알파벳에 대한 색상을 이해할 수 있지만, 이는 피그마를 확인하는 과정을 거치기 때문에 비효율적이라고 생각합니다. 그래서 tailwind의 느낌으로 색상명을 명확하게 작성해주고 숫자로 밝기 수준을 나타내는 방법도 좋을 것 같습니다.

Legitgoons replied:
향후 유지 보수 측면까지 생각한다면 색상명을 정확하게 적는게 더 이점이 있을 수 있겠네요. 수정하겠습니다!

BangDori commented on src/shared/styles/fonts.scss:
```diff
@@ -0,0 +1,31 @@
+.b2s { @include font-style(semibold, 0.6875, -3%, 130%); }
+.b2r { @include font-style(regular, 0.6875, -3%, 130%); }
+.b3m { @include font-style(medium, 0.5625, -3%, 130%); }
+.b4m { @include font-style(medium, 0.5, -3%, 130%); }
```
P4: 이 부분도 네이밍을 이렇게 하신 이유가 궁금합니다!

Legitgoons replied:
font class는 font size + weight 앞 글자로 만들었습니다. 
- font size: 디자인 팀에서 명명한 사이즈를 그대로 사용하는 것이 마찬가지로 작업 효율면에서 나을 것이라고 생각했습니다.
- weight: bold, semibold, regular, medium 4가지 경우만 있기에 앞 글자만 사용해도 헷갈리지 않을 것이라 생각했습니다.

고려한 방안: weight + size number를 사용하는 방안(예시: b24, s24, s20, s16...)을 고려했습니다.
- 그러나 추후 사이즈가 변경된다면 사용한 모든 파일에서 클래스명을 변경해줘야 한다는 단점이 있어 선택하지 않았습니다.

BangDori replied:
1. 이 부분도 폰트 클래스명을 조금 더 명확하게 하면 좋을 것 같습니다!
2. 그리고, `$spacing` 파라미터는 `-3%`만 받고 있는데 굳이 필요하지 않나요?

근데, Mixin을 활용해서 font-style을 함수처럼 활용하는 방법이 근데 정말 멋있네요. 의찬님 scss 처음 아니시죠..?

Legitgoons replied:
1. 클래스명을 좀 더 명확하게 하려면 font name / font size / weight를 추가적으로 작성해줘야 할 듯 한데, font name은 모두 동일 font를 사용하기에 의미가 없고, size는 만약 디자인팀에서 변경 요청 들어오면 난감해지니 weight를 추가적으로 작성하는 식으로 해보겠습니다.

2. 맞네용 ㅎㅎ;; $spacing 제거하겠습니다.

BangDori reviewed (changes requested):
Navbar 컴포넌트와 스타일 시트의 위치가 FSD Architecture가 지향하는 구조와 조금 맞지 않다고 생각하여 Request Changes 하였습니다! 확인 부탁드립니다 ㅎㅎ

Legitgoons committed:
feat: FNB 컴포넌트와 스타일 파일 이동 및 이름 변경

shared/ui에 많은 컴포넌트가 들어갈 수 있기에 추가적으로 depth를 만들었습니다.
상단 navbar와 헷갈릴 여지가 있기에 FNB로 이름을 변경했습니다.

Legitgoons committed:
feat: RootLayout FNB 컴포넌트 이름 변경 적용

Legitgoons committed:
feat: FNB.scss 파일 명 naming rule에 따라 수정

Legitgoons committed:
feat: 합의에 따라 colors 변수명 수정

Legitgoons committed:
feat: font naming 수정 및 font-style  삭제

BangDori commented on src/shared/ui/FNB/FNB.tsx:
P1: 현재 FNB 컴포넌트에서 불러오고 있는 스타일 시트의 경로가 잘못되어 AWS Amplify에서 오류가 발생하고 있는 것 같습니다!

Legitgoons replied:
헉; 바로 수정했습니다

BangDori reviewed (changes requested):
마지막으로 하나만 확인 부탁드립니다!

Legitgoons committed:
fix: FNB.scss 경로 수정

BangDori reviewed (approved):
고생하셨습니다! 승인하면 될 것 같아요.
